lib/TPP/ConvertPerfToFunc.cpp

```diff
@@ -0,0 +1,386 @@
+
+#include "TPP/Dialect/Perf/PerfOps.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::perf;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Perf dialect type normalization helper function.
+// Cast memref and tensor to their unranked versions and convert
+// custom perf types into default primitive types.
+// Leave all the other operands as they are.
+static SmallVector<Type> extractNormalizedTypes(OpBuilder &b,
+                                                ValueRange values) {
+  SmallVector<Type> results;
+  results.reserve(values.size());
+
+  for (Value val : values) {
+    TypeSwitch<Type>(val.getType())
+        .Case<MemRefType>([&](Type t) {
+          auto memrefType = cast<MemRefType>(t);
+          auto unrankedMemref = UnrankedMemRefType::get(
+              memrefType.getElementType(), memrefType.getMemorySpace());
+          results.push_back(unrankedMemref);
+        })
+        .Case<TensorType>([&](Type t) {
+          auto tensorType = cast<TensorType>(t);
+          auto unrankedTensor =
+              UnrankedTensorType::get(tensorType.getElementType());
+          results.push_back(unrankedTensor);
+        })
+        .Case<TimerType>([&](Type t) {
+          auto i64 = IntegerType::get(b.getContext(), 64);
+          results.push_back(i64);
+        })
+        .Default([&](Type t) { results.push_back(t); });
+  }
+
+  return results;
+}
+
+// Similar to 'extractNormalizedTypes' but inserts conversions from original
+// types to normalized ones when possible.
+// The conversions aim to simplify runtime function signature generation by, for
+// example, erasing explicit memref and tensor shapes.
+static SmallVector<Value> getNormalizedOperands(OpBuilder &b, Location loc,
+                                                ValueRange operands) {
+  SmallVector<Value> res;
+  res.reserve(operands.size());
+
+  for (Value op : operands) {
+    auto type = extractNormalizedTypes(b, op);
+    TypeSwitch<Type>(op.getType())
+        .Case<MemRefType>([&](Type t) {
+          Value cast = b.create<memref::CastOp>(loc, type, op);
+          res.push_back(cast);
+        })
+        .Case<TensorType>([&](Type t) {
+          Value cast = b.create<tensor::CastOp>(loc, type, op);
+          res.push_back(cast);
+        })
+        .Default([&](Type t) { res.push_back(op); });
+  }
+
+  return res;
+}
+
+// Create a perf function prototype.
+// Assumes that the current insertion point is in the correct place.
```

Comment:
This assumption is dangerous because functions can only be inserted into modules.

You also have two levels of `InsertionGuard` (main loop and `buildPerf...`) which are redundant.

Technically, you only need one, and change the insertion point to the module at the `buildPerf...` level, after creating the `InsertionGuard`, and it should all work, without assumptions.

Reply:
True, these functions should be more independent.
Double guards introduce a bit of redundancy but it should be safer to have them whenever the insertion point is changed. Then the functional call inserter can just blindly call the builders.